bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/TokenResource.java

```diff
@@ -0,0 +1,337 @@
+import javax.ws.rs.core.Response.ResponseBuilder;
+import javax.ws.rs.core.Response.Status;
+import javax.ws.rs.core.SecurityContext;
+import javax.ws.rs.core.UriInfo;
+
+import org.jose4j.base64url.Base64Url;
+import org.openhab.core.auth.AuthenticationProvider;
+import org.openhab.core.auth.ManagedUser;
+import org.openhab.core.auth.PendingToken;
+import org.openhab.core.auth.User;
+import org.openhab.core.auth.UserRegistry;
+import org.openhab.core.auth.UserSession;
+import org.openhab.core.io.rest.RESTResource;
+import org.openhab.core.io.rest.Stream2JSONInputStream;
+import org.openhab.core.io.rest.auth.internal.TokenEndpointException.ErrorType;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+
+/**
+ * This class is used to issue JWT tokens to clients.
+ *
+ * @author Yannick Schaus - Initial contribution
+ */
+@Path(TokenResource.PATH_AUTH)
+@Api(value = TokenResource.PATH_AUTH)
+@Component(service = { RESTResource.class, TokenResource.class })
+public class TokenResource implements RESTResource {
+    private final Logger logger = LoggerFactory.getLogger(TokenResource.class);
+
+    /** The URI path to this resource */
+    public static final String PATH_AUTH = "auth";
+
+    /** The name of the HTTP-only cookie holding the session ID */
+    public static final String SESSIONID_COOKIE_NAME = "X-OPENHAB-SESSIONID";
+
+    @Context
+    private UriInfo uriInfo;
+
+    @Reference
+    private AuthenticationProvider authenticationProvider;
+
+    private UserRegistry userRegistry;
+
+    private JwtHelper jwtHelper;
+
+    @Activate
+    public TokenResource(final @Reference UserRegistry userRegistry, final @Reference JwtHelper jwtHelper) {
+        this.userRegistry = userRegistry;
+        this.jwtHelper = jwtHelper;
+    }
+
+    @POST
+    @Path("/token")
+    @Produces({ MediaType.APPLICATION_JSON })
+    @Consumes({ MediaType.APPLICATION_FORM_URLENCODED })
+    @ApiOperation(value = "Get access and refresh tokens.")
+    @ApiResponses(value = { @ApiResponse(code = 200, message = "OK") })
+    public Response getToken(@FormParam("grant_type") String grantType, @FormParam("code") String code,
+            @FormParam("redirect_uri") String redirectUri, @FormParam("client_id") String clientId,
+            @FormParam("refresh_token") String refreshToken, @FormParam("code_verifier") String codeVerifier,
+            @QueryParam("useCookie") boolean useCookie, @CookieParam(SESSIONID_COOKIE_NAME) Cookie sessionCookie) {
+        try {
+            switch (grantType) {
+                case "authorization_code":
+                    // find an user with the authorization code pending
```

Comment:
Hell, these are many lines of code for this case of the switch statement - it would be much nicer to move this codeblock to a separate method.